File: ueb6/client/src/main/scala/Messages.scala
import io.circe.{Decoder, DecodingFailure, Encoder, Json}
import io.circe.syntax._

sealed trait ClientMessage
case class ClientEdit(
  rev: Int,
  op: TextOperation.Operation,
  cursor: Int) extends ClientMessage

sealed trait ServerMessage
case object Ack extends ServerMessage
case object HelloGuys extends ServerMessage 
case class RemoteEdit(
  op: TextOperation.Operation,
  cursors: List[Int]
) extends ServerMessage

object ClientMessage {
  implicit val encodeAction =
    Encoder.instance[TextAction] {
      case Retain => 1.asJson
      case Insert(c) => c.asJson
      case Delete => (-1).asJson
    }

  implicit val encodeClientMessage = Encoder.instance[ClientMessage] {
    case ClientEdit(rev,op,cursor) => Json.obj(
      "rev" -> rev.asJson,
      "op" -> op.asJson,
      "cursor" -> cursor.asJson
    )
  }
}

object ServerMessage {
  implicit val parseAction =
    Decoder.decodeChar.map[TextAction](Insert) or
    Decoder.decodeInt.emap[TextAction] {
      case -1 => Right(Delete)
      case 1 => Right(Retain)
      case other => Left("invalid text action")
    }

  val parseAckAndHello = Decoder.decodeString.emap[ServerMessage] {
    case "ack" => Right(Ack)
    case "helloguys" => Right(HelloGuys)
    case other => Left("invalid server message")
  }


  val parseRemoteEdit : Decoder[ServerMessage] = 
    Decoder.instance( o =>
      for {
        op <- o.downField("op").as[List[TextAction]] 
        cursors <- o.downField("cursors").as[List[Int]]
      } yield RemoteEdit(op,cursors)
    )

  implicit val parseServerMessage =
    parseAckAndHello or parseRemoteEdit
}

File: ueb6/app/Messages.hs
{-# LANGUAGE OverloadedStrings #-}

module Messages where

import qualified Data.Map.Lazy as M
import Data.Aeson
import OT

data ClientMessage = ClientEdit Int TextOperation Cursor

data ServerMessage = RemoteEdit TextOperation (M.Map Int Cursor)
                   | Ack
                   | HelloGuys

instance FromJSON TextAction where
  parseJSON (Number x) = do
    n <- parseJSON (Number x)
    return $ 
      if n < 0 then Delete $ abs n
      else Retain n 
  parseJSON (String t) = do
    c <- parseJSON (String t)
    return $ Insert c

instance FromJSON TextOperation where
  parseJSON (Array a) = do
    ops <- parseJSON (Array a)
    return $ TextOperation ops

instance FromJSON ClientMessage where
  parseJSON (Object o) = ClientEdit
    <$> o .: "rev"
    <*> o .: "op"
    <*> o .: "cursor" 

instance ToJSON TextAction where
  toJSON (Retain n) = toJSON (n :: Int)
  toJSON (Delete n) = toJSON (-n :: Int)
  toJSON (Insert c) = toJSON c

instance ToJSON TextOperation where
  toJSON (TextOperation as) = toJSON $ map toJSON as

instance ToJSON ServerMessage where
  toJSON Ack = toJSON ("ack" :: String)
  toJSON HelloGuys = toJSON ("helloguys" :: String)
  toJSON (RemoteEdit op cursors) = object ["op" .= toJSON op, 
    "cursors" .= cursors'] where cursors' = toJSON $ M.elems cursors 
 
